kvdb/etcd/readwrite_bucket.go

```diff
@@ -66,6 +70,29 @@ func (b *readWriteBucket) ForEach(cb func(k, v []byte) error) error {
 	return nil
 }
 
+// ForAll is an optimized version of ForEach with the limitation that no
+// additional queries can be executed within the callback.
+func (b *readWriteBucket) ForAll(cb func(k, v []byte) error) error {
+	// When we opened this bucket, we fetched the bucket key using the STM
+	// which put a revision "lock" in the read set. We can leverage this
+	// by incrementing the revision on the bucket, making any transaction
+	// retry that'd touch this same bucket. This way we can safely read all
+	// keys from the bucket and not cache them in the STM.
+	// To increment the bucket's revision, we simply put in the bucket key
+	// value again (which is idempotent if the bucket has just been created).
+	b.tx.stm.Put(string(b.key), string(b.id))
```

Comment:
Not sure I fully understand the magic performed here. If I read the comment correctly, this both prevents interfering with the STM cache that we'd have for normal calls as well as make sure we retry the full `ForAll` if anything within this bucket changes?

Reply:
Yes, but it has a price that concurrent access will result in retries if some other txn were to access the parent bucket. Fortunately in this specific case where we use `ForAll` it's not an issue.